internal/pkg/source/source.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package source contains the logic related to the concept of the source which may be tainted.
+package source
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/eapache/queue"
+	"github.com/google/go-flow-levee/internal/pkg/sanitizer"
+	"golang.org/x/tools/go/ssa"
+)
+
+type classifier interface {
+	IsSanitizer(*ssa.Call) bool
+	IsPropagator(*ssa.Call) bool
+	IsSourceFieldAddr(*ssa.FieldAddr) bool
+}
+
+// Source represents a Source in an SSA call tree.
+// It is based on ssa.Node, with the added functionality of computing the recursive graph of
+// its referrers.
+// Source.sanitized notes sanitizer calls that sanitize this Source
+type Source struct {
+	node       ssa.Node
+	marked     map[ssa.Node]bool
+	sanitizers []*sanitizer.Sanitizer
+	config     classifier
+}
+
+// Node returns the underlying ssa.Node of the Source.
+func (a *Source) Node() ssa.Node {
```

Comment:
I agree. 
I will fix this on a separate PR.

Reply:
In general, I agree.
However, since a Source could be referenced by Values and Instructions, it forces us to use ssa.Node to keep track of vertices, right?

Reply:
I'm not sure that's the case.  Generally, a source will be sanitize/propagated/sunk by an `ssa.Instruction`

The following are the only `ssa` types which implement `Value` but not `Instruction`.

```
*Builtin
*Const
*FreeVar
*Function
*Global
*Parameter
```

So long as those types are part of source identification, I don't think we'll lose coverage.  But it would require another pass at the `bfs` / dominance logic, so probably shouldn't be part of this PR.

Reply:
I think we are on the same page.
I agree that the constructor of the source should accept `ssa.Value` and not `ssa.Node`. 

However, the graph of the source will contain Instructions (example `ssa.Store`), and since the `Source` itself is part of the graph, vertices would need to support a broader interface - probably Node.
Let's pick this up on a separate PR.

Reply:
> Let's pick this up on a separate PR.

+1.